### PR TITLE
Cleaned up and implemented name conflict resolving

### PR DESCRIPTION
**Pull request by ilyakooo0**
Some minor code refactoring
added support for building with stack
Added ability to resolve field name conflicts

NOTE: doesn't currently work with the master branch of xml-isogen, requires this branch to be merged: https://github.com/typeable/xml-isogen/tree/digits

### CODE REVIEW

**Comment by ilyakooo0** [comment]
TODO: 

- [x] CLI argument for header
- [ ] Extract some sort of header comment "API blah blah blah 14.2"

**Comment by ilyakooo0** [comment]
The TODO should probably be done in separate pull requests.

**Comment by dredozubov** [comment]
> NOTE: doesn't currently work with the master branch of xml-isogen, requires this branch to be merged: https://github.com/typeable/xml-isogen/tree/digits

Can you please merge it into this PR?

**Comment by ilyakooo0** [comment]
@dredozubov I don't have write access

**Comment by ilyakooo0** [comment]
UPD: CLI arguments have been implemented in #3

File: src/Main.hs [hunk 1]
```diff
@@ -219,22 +278,22 @@ quote :: Text -> Text
...
 
 -- | Nothing means unknown name
-typeName :: DatatypeRef -> Maybe Text
-typeName (InlineComplex dt) = result
-  where
-    result = case dt of
-      TypeComplex (ComplexType n _ _ _) -> n
-      TypeSimple (STAtomic _ sat _ _)     -> Just $ simpleTypeToIsoType QQ sat
-typeName (DatatypeRef t)               = Just t
+-- typeName :: DatatypeRef -> Maybe Text
```
Comment (dredozubov) [inline src/Main.hs:281]: Can you remove dead code here?

File: src/Main.hs [hunk 2]
```diff
@@ -264,79 +323,99 @@ wrap80 = L.foldl' go [] . T.split (==' ')
...
-  (TypeComplex (ComplexType iName attrs annotations mGroupSchema))
+  (TypeComplex (ComplexType _ _ tyAnnotations mGroupSchema))
   annotationsRec = do
-    name <- case eName of
-      Just name -> pure name
-      Nothing   -> case iName of
-        Just name' -> pure name'
-        Nothing    -> error $ "[generateIsoDatatype] no name was provided"
+    let finalName = eName
+    -- finalName <- case eName of
```
Comment (dredozubov) [inline src/Main.hs:343]: dead code

File: src/Main.hs [hunk 3]
```diff
@@ -264,79 +323,99 @@ wrap80 = L.foldl' go [] . T.split (==' ')
...
   genType
   eName
-  (TypeSimple (STAtomic iName sat annotations restrictions))
+  (TypeSimple ty@(STAtomic iName sat tyAnnotations restrictions))
   annotationsRec = do
-    name <- case eName of
-      Just name -> pure name
-      Nothing   -> pure iName
+    let finalName = eName
+    -- finalName <- case eName of
```
Comment (dredozubov) [inline src/Main.hs:370]: dead code

File: src/Main.hs [hunk 4]
```diff
@@ -264,79 +323,99 @@ wrap80 = L.foldl' go [] . T.split (==' ')
...
         toComment Haddock $ annotationsRec
           <> [Documentation ""]
-          <> annotations
-      isEnumeration (Enumeration _)   = True
-      fromEnumeration (Enumeration x) = x
-      enums                           = Prelude.filter isEnumeration restrictions
-      typeStr                         = simpleTypeToIsoType QQ sat
-    if Prelude.length enums > 0
+          <> tyAnnotations
+      -- typeStr                         = simpleTypeToIsoType QQ sat
```
Comment (dredozubov) [inline src/Main.hs:378]: commented code

File: src/Main.hs [hunk 5]
```diff
@@ -74,11 +105,40 @@ proceed' el = do
...
     (_, e') = execRWS sortElemsByDeps (M.fromList dList) e
 
+type Abbreviation = Text
+type Subelement = Text
+type DeconflicterMonad = State (M.Map Abbreviation (S.Set Subelement))
+
+deconflictNames :: [(Text, Datatype)] -> [(Text, Datatype)]
+deconflictNames = flip evalState mempty . mapM deconflictName
+
+deconflictName :: (Text, Datatype) -> DeconflicterMonad (Text, Datatype)
```
Comment (dredozubov) [inline src/Main.hs:120]: Probably should be called `disambiguate` instead. Please add a haddock describing the problem and solution here.

File: README.md [hunk 6]
```diff
@@ -10,8 +10,3 @@ Use `xsd-isogen in.xsd Generator > out.hs` for queries.
 Use `xsd-isogen in.xsd Parser > out.hs` for responses.
 
 Use `xsd-isogen in.xsd Both > out.hs` for common definitions.
-
-Keep in mind that name clashes are possible and should be disambiguated
```
Comment (dredozubov) [inline README.md:14]: Should be substituted with a description of disambiguation renaming process.

File: src/Main.hs [hunk 7]
```diff
@@ -349,54 +428,56 @@ generateEnum
...
-lookupSimpleBaseType
-  :: DatatypeRef
-  -> GenMonad (Maybe SimpleAtomicType)
-lookupSimpleBaseType (InlineComplex  _) = pure Nothing
-lookupSimpleBaseType (DatatypeRef tyName) = do
-  M.lookup tyName <$> ask >>= \case
-    Just (TypeComplex _)                 -> pure Nothing
-    Just (TypeSimple (STAtomic _ sat _ _)) -> pure $ Just sat
-    Nothing                              -> error $ "can't look up type: " <> show tyName
+-- lookupSimpleBaseType
```
Comment (dredozubov) [inline src/Main.hs:438]: Why was it commented? It should either be removed or kept as is.

File: src/Main.hs [hunk 8]
```diff
@@ -120,12 +180,11 @@ lookupDatatype :: Text -> PreMonad ()
...
   traceM $ "lookupDatatype: " <> show ty
   (dm, s) <- get
-  let mTy = M.lookup ty dm
   case M.lookup ty dm of
-    Nothing -> if S.member ty s then pure () else do
+    Nothing -> if S.member ty s then pure () else
       error $ "can't look up datatype: " <> show ty
     Just typ -> case references typ of
-      [] -> proceed ty typ >> sortDatatypesByDeps
+      [] -> proceed ty typ -- >> sortDatatypesByDeps -- Pretty sure this was a bug
```
Comment (dredozubov) [inline src/Main.hs:187]: dead piece of code

File: xsd-isogen.cabal [hunk 9]
```diff
@@ -32,3 +32,4 @@ executable xsd-isogen
                      , unordered-containers
   hs-source-dirs:      src
   default-language:    Haskell2010
+  ghc-options: -fwarn-unused-binds -fwarn-unused-imports -Wall -Wincomplete-uni-patterns -Wincomplete-record-updates -Wpartial-fields -Werror=missing-home-modules -Wmissing-home-modules -Widentities -Wredundant-constraints -Wmissing-export-lists
```
Comment (dredozubov) [inline xsd-isogen.cabal:35]: Can you turn it into a vertical list?

File: xsd-isogen.cabal [hunk 10]
```diff
@@ -10,12 +10,12 @@ synopsis:            Generate xml-isogen DSL from xsd files
 homepage:            github.com/dredozubov/xsd-isogen
 -- bug-reports:
 license:             BSD-3-Clause
-license-file:        LICENSE
+-- license-file:        LICENSE
```
Comment (dredozubov) [inline xsd-isogen.cabal:13]: let's add BSD-3 license instead if cabal gives an error during the build

File: xsd-isogen.cabal [hunk 11]
```diff
@@ -10,12 +10,12 @@ synopsis:            Generate xml-isogen DSL from xsd files
...
 author:              Denis Redozubov
 maintainer:          denis.redozubov@gmail.com
 -- copyright:
 category:            Text
-extra-source-files:  CHANGELOG.md
+-- extra-source-files:  CHANGELOG.md
```
Comment (dredozubov) [inline xsd-isogen.cabal:18]: Let's remove it altogether or add a changelog

File: stack.yaml [hunk 12]
```diff
@@ -0,0 +1,69 @@
+# This file was automatically generated by 'stack init'
```
Comment (dredozubov) [inline stack.yaml:1]: I want to keep cabal-install as a primary build tool for this repo. Do we really need it here?

**Review by dredozubov — CHANGES_REQUESTED** [review]
I've left some comments and I'd like to ask you to add a description to this PR, so other people that haven't had discussion we had could review it. Also, I think it would be very beneficial to separate this into a few PRs.

File: README.md (hunk 6 above)
Comment (ilyakooo0) [reply README.md:14]: Done in #3

**Comment by ilyakooo0** [comment]
Decomposed into #4 and #5

**Pull request closed without merging**